core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java

```diff
@@ -109,50 +112,58 @@ private void setPrimariesInitialRecoveries(int primariesInitialRecoveries) {
 
     @Override
     public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
-        if (shardRouting.primary()) {
-            assert shardRouting.unassigned() || shardRouting.active();
-            if (shardRouting.unassigned()) {
-                // primary is unassigned, means we are going to do recovery from gateway
-                // count *just the primary* currently doing recovery on the node and check against concurrent_recoveries
-                int primariesInRecovery = 0;
-                for (ShardRouting shard : node) {
-                    // when a primary shard is INITIALIZING, it can be because of *initial recovery* or *relocation from another node*
-                    // we only count initial recoveries here, so we need to make sure that relocating node is null
-                    if (shard.initializing() && shard.primary() && shard.relocatingNodeId() == null) {
-                        primariesInRecovery++;
-                    }
+        assert shardRouting.primary() == false || shardRouting.unassigned() || shardRouting.active();
+
+        if (shardRouting.primary() == false || shardRouting.unassigned() == false) {
```

Comment:
Unfortunately we cannot use `shardrouting.isPeerRecovery()` here because the `shardRouting` that is passed to the method is not the initializing shard that will be assigned to this node. I will add some more comments.

I can also invert the flow of this `if/else` condition so that we have `shardRouting.primary() && shardRouting.unassigned()` as boolean condition which corresponds to what we currently have in the `else` branch.